transformations/p1_noun_transformation/transformation.py

```diff
@@ -0,0 +1,123 @@
+import json
+import os
+import sys
+import nltk
+
+from TestRunner import convert_to_snake_case
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+import wptools
+from bs4 import BeautifulSoup
+from nltk.corpus import stopwords
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+def get_cd(phrase):
+    sys.stdout = open(os.devnull, "w")
+    sys.stderr = open(os.devnull, "w")
+    so = wptools.page(phrase, silent=True).get_parse()
+    sys.stdout = sys.__stdout__
+    sys.stderr = sys.__stderr__
+    a = so.data["parsetree"]
+    short_desc = BeautifulSoup(a, 'html.parser').find('title').nextSibling.contents[1].contents[0]
+    return short_desc
+
+
+def stop_word_removal(text_all, cached_stop_words):
+    new_text_all = []
+    for text in text_all:
+        text1 = ' '.join([word for word in text.split() if word not in cached_stop_words])
+        new_text_all.append(text1)
+    return new_text_all
+
+
+def get_grams(sentence, N):
+    all_grams = []
+    for l in range(1, N + 1):
+        grams = [" ".join(sentence[i:i + l]) for i in range(len(sentence) - l + 1)]
+        all_grams = all_grams + grams
+    return all_grams
+
+
+def prep_sent(sentence, n, cached_stop_words):
+    list1 = list(set(stop_word_removal(get_grams(sentence.split(), n), cached_stop_words)))
+    list1 = list(filter(None, list1))
+    list1 = list(sorted(list1, key=len))[::-1]
+    return list1
+
+
+def get_noun_definitions(inp_sent, cached_stop_words):
+    max_ngrams = 3
+    inp_sent = inp_sent.lower()
+    tokens = nltk.word_tokenize(inp_sent)
+    pos_tagged = nltk.pos_tag(tokens)
+
+    non_noun_tuples = list(filter(lambda x: x[1][0] != 'N', pos_tagged))
+    non_nouns = [a_tuple[0] for a_tuple in non_noun_tuples]
+
+    mentions = prep_sent(inp_sent, max_ngrams, cached_stop_words)
+    mentions_dict = {}
+    done_grams = non_nouns
+    for men in mentions:
+        if men in done_grams:
+            continue
+        val = None
+        try:
+            val = get_cd(men)
+        except:
+            continue
+        if val is not None:
+            mentions_dict[men] = val
+            all_to_stop = get_grams(men.split(), len(men.split()))
+            done_grams = done_grams + all_to_stop
+    for key_phrase, phrase_definition in mentions_dict.items():
+        if len(phrase_definition.split()) == 1:
+            continue
+        inp_sent = inp_sent.replace(key_phrase, key_phrase + "(" + phrase_definition.lower() + ")")
+    return (inp_sent)
+
+
+class AddNounDefinition(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.TEXT_TAGGING,
```

Comment:
Please remove TaskType.TEXT_TAGGING, as this transformation, can not be used for tagging tasks (due to difference in the number of words in generated transformation)

Reply:
Done